src/test/java/com/atlassian/db/replica/api/TestDualConnection.java

```diff
@@ -690,7 +701,7 @@ public void shouldExecuteReadOperationOnReplica() throws SQLException {
             .containsOnly(REPLICA);
         verify(databaseCall).call(
             any(),
-            eq(new RouteDecisionBuilder(Reason.READ_OPERATION).sql(SIMPLE_QUERY).build())
+            eq(new RouteDecisionBuilder(Reason.READ_OPERATION).sql(SIMPLE_QUERY).isSqlWrite(false).build())
```

Comment:
Now, READ_OPERATION sqlWrite=true is possible 😢 . Why can't we keep reducing the states at the compile time?

Reply:
@wyrzyk Can you explain ?
READ_OPERATION -> means that always sqlWrite=false.
 
I'd like to reduce state at the compile time. How I can achieve that?

Reply:
What will happen in case I use the API this way?
```
new RouteDecisionBuilder(Reason.READ_OPERATION).sql(SIMPLE_QUERY).isSqlWrite(true).build()
```

Reply:
aaaa I think I know what you meant!
I've tried hard to show that Reasons doesn't reflect unambiguously isWrite, but I failed and looked like they reflect, at least considering what we know right know and that we're aware of https://github.com/atlassian-labs/db-replica/issues/53
So actually I can continue with 1st idea to add isWrite to Reason!

`    private final String name;
    private final boolean isRunOnMain;
    private final boolean isWrite;
    

    private Reason(final String name, boolean isRunOnMain, boolean isWrite) {
        this.name = name;
        this.isRunOnMain = isRunOnMain;
        this.isWrite = isWrite;
    }

    public static final Reason RW_API_CALL = new Reason("RW_API_CALL", true, true);
    public static final Reason REPLICA_INCONSISTENT = new Reason("REPLICA_INCONSISTENT", true, false);
    public static final Reason READ_OPERATION = new Reason("READ_OPERATION", false, false);
    public static final Reason WRITE_OPERATION = new Reason("WRITE_OPERATION", true, true);
    public static final Reason LOCK = new Reason("LOCK", true, true);
    public static final Reason MAIN_CONNECTION_REUSE = new Reason(
        "MAIN_CONNECTION_REUSE",
        true,
        false);
    public static final Reason HIGH_TRANSACTION_ISOLATION_LEVEL = new Reason(
        "HIGH_TRANSACTION_ISOLATION_LEVEL",
        true,
        false);
    public static final Reason RO_API_CALL = new Reason("RO_API_CALL", false, false);
`

Reply:
> I'd like to reduce state at the compile time. How I can achieve that?

In the same way as `isRunOnMain` it would have to be mapped 1:1 with reasons. We may need a second class to map it.

Reply:
Or 2nd idea
IsWrite()
reason.equals ( RW_API_CALL | WRITE_OPERATION | LOCK ) 

 :) 
this might be simpler but it left ambiguity in the code for further devs.
Thats why I'm for 1st option.

Reply:
OTOH, Maciej wanted to bind query with `writesToDatabase`, this way we can't make it compile time as queries are not known at this time 🤔 .

Reply:
If not possible to make it OO and compile-time safe, we can introduce runtime checks.

Reply:
I like both approaches.

Reply:
A `Reason` (or a whole `RouteDecision`) can be inferred **from** `SqlQuery` rather than independently. OO✔️ compile-time✔️

Reply:
Good discusion! 👍 

> A Reason (or a whole RouteDecision) can be inferred from SqlQuery 
Hmm but where is the information whether Sql was invoked as writeStatement or readStatement which is pre requisite to determine `isWrite` or Reason itself.

I think I can make it OOP and compile time but SQLQuery would have to have information about read/write statement as we don't parse SQL itself and this is our source of knowledge.
Or create SqlQuery with boolean .

@dagguh  How you see it both OO and compile-time ?

Reply:
I'll go with

```
public class SqlQuery {

    private final String query;
    private final boolean isWriteStatement;
    
    public SqlQuery(String query, boolean isWriteStatement) {
        this.query = query;
        this.isWriteStatement = isWriteStatement;
    }

    public Boolean writesToDatabase() {
```

Reply:
Give me 15 mins and there will be PR, I think you will be happy with it.

Reply:
`SqlQuery` doesn't need to know about "the context" like `boolean isWriteStatement`. It's just a query.
I said that **a** (some) reason can be inferred, rather than **all** reasons.

Reply:
Concretely we could ask `SqlQuery` for `RouteDecision route()`.

Reply:
The problem is that over there RouteDecision isn't known, and rather it's established overthere, what worse, reason can change later, down in the code in when ConnectionState is determined as RouteDecision is derivative from SQL + statement + connection state (and order of those checks too).

RouteDecision is built just before doing execute
```
return execute(statement::execute, decisionBuilder.build());
```


I'll try to somehow make it OOP, and compileTime + runtime. Either SqlQuery need to have SqlQueryBuilder, and rather RouteDecision will contain SqlQuery (as it currently contain sql) .

I'm not sure whether SqlQuery should be added to the API 🤔 

I like your thinking @dagguh, but looks like you missed some of impl detail which blocks the implementation of it.
once again  RouteDecision is derivative from SQL + statement + connection state

Another broken part is that SQL function maybe customized: ``SqlFunction(Set<String> readOnlyFunctions)`` and this need to come from DualConnectionBuilder.
That's why SQLQuery  goes together with Reason, not independently

Reply:
I've pushed changes wtih compile time explicit isWrite information on Reason lvl. I also introduce SqlQuery to increase OO. 
It's not exactly the same because of mentioned reasons(mainly `RouteDecision` is build later, and `RouteDecisionBuilder` is used not only for Statements(SQL) but for choosing connection(based on connection state) - in those places adding `SqlQuery` doesn't make sense.

This might be improved in future but for now it's +1 improvement into good direction.

Reply:
Just try it

Reply:
Tried, I hope it met your expectations?

Reply:
Do you have it on a side branch?

Reply:
Each `Reason.isWrite` is statically predetermined now, so inferring it from `SqlQuery` is no longer needed for compile-time invariant 👍